sessiontxn/isolation/serializable.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package isolation
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/sessiontxn"
+)
+
+// PessimisticSerializableTxnContextProvider provides txn context for isolation level oracle-like serializable
+type PessimisticSerializableTxnContextProvider struct {
+	baseTxnContextProvider
+}
+
+// NewPessimisticSerializableTxnContextProvider returns a new PessimisticSerializableTxnContextProvider
+func NewPessimisticSerializableTxnContextProvider(sctx sessionctx.Context,
+	causalConsistencyOnly bool) *PessimisticSerializableTxnContextProvider {
+	provider := &PessimisticSerializableTxnContextProvider{
+		baseTxnContextProvider{
+			sctx:                  sctx,
+			causalConsistencyOnly: causalConsistencyOnly,
+			onInitializeTxnCtx: func(txnCtx *variable.TransactionContext) {
+				txnCtx.IsPessimistic = true
+				txnCtx.Isolation = ast.Serializable
+			},
+			onTxnActive: func(txn kv.Transaction) {
+				txn.SetOption(kv.Pessimistic, true)
+			},
+		},
+	}
+
+	provider.getStmtForUpdateTSFunc = provider.getTxnStartTS
+	provider.getStmtReadTSFunc = provider.getTxnStartTS
+	return provider
+}
+
+// OnStmtRetry is the hook that should be called when a statement is retried internally.
+func (p *PessimisticSerializableTxnContextProvider) OnStmtRetry(_ context.Context) error {
+	return errors.New("Should never retry in serializable isolation")
```

Comment:
I think we can allow stmt retry here. An assumption that it will not retry on serializable isolation is a little arbitrary because the statement can retry for other non-transaction reasons theoretically.

Reply:
Done